src/magic_folder/status.py

```diff
@@ -0,0 +1,202 @@
+)
+
+from twisted.application import (
+    service,
+)
+
+
+class IStatus(Interface):
+    """
+    An internal API for services to report realtime status
+    information. These don't necessarily correspond 1:1 to outgoing
+    messages from the status API.
+    """
+
+    def upload_started():
+        """
+        One or more items are now in our upload queue.
+        """
+
+    def upload_stopped():
+        """
+        No items are in the upload queue.
+        """
+
+
+class StatusProtocol(WebSocketServerProtocol):
+    """
+    Speaks the server side of the WebSocket status protocol, usually
+    mounted at /v1/status from our web API.
+
+    This is authenticated with the same Bearer token as the rest of
+    the /v1 API.
+    """
+
+    def __init__(self, status):
+        self.status = status
+        WebSocketServerProtocol.__init__(self)
+
+    def onOpen(self):
+        """
+        WebSocket API: successful handshake
+        """
+        self.status.client_connected(self)
+
+    def onClose(self, wasClean, code, reason):
+        """
+        WebSocket API: we've lost our connection for some reason
+        """
+        self.status.client_disconnected(self)
+
+    def onMessage(self, payload, isBinary):
+        """
+        WebSocket API: a message has been received from the client. This
+        should never happen in our protocol.
+        """
+        self.sendClose(
+            code=3000,
+            reason="Unexpected incoming message",
+        )
+
+
+class StatusFactory(WebSocketServerFactory):
+    """
+    Instantiates server-side StatusProtocol instances when clients
+    connect.
+    """
+    protocol = StatusProtocol
+
+    def __init__(self, status):
+        """
+        :param WebSocketStatusService status: actual provider of our
+            status information. The protocol will use this to track
+            clients as they connect and disconnect.
+        """
+        self._status = status
+        WebSocketServerFactory.__init__(self, server="magic-folder")
+
+    def buildProtocol(self, addr):
+        """
+        IFactory API
+        """
+        protocol = self.protocol(self._status)
+        protocol.factory = self
+        return protocol
+
+
+@attr.s
+@implementer(service.IService)
+@implementer(IStatus)
+class WebSocketStatusService(service.Service):
+    """
+    A global service that can be used to report status information via
+    an authenticated WebSocket connection.
+
+    The authentication mechanism is the same as for the HTTP API (see
+    web.py where a WebSocketResource is mounted into the resource
+    tree).
+    """
+
+    # tracks currently-connected clients
+    _clients = attr.ib(default=attr.Factory(set))
+
+    # the last state we marshaled. This is the last state we sent out
+    # and any newly connecting client will receive it immediately.
+    _last_state = attr.ib(default=None)
+
+    # current live state
+    _uploading = attr.ib(default=False)
```

Comment:
I stored the last "actual thing we sent" to have maximum consistency between clients...what are you suggesting storing, exactly?

Reply:
(I'm not sure what you mean by the no-op state updates comment?)

Reply:
The idea here is that we:
- update the state
- compare it to our last state
- if different:
  - send this state to clients

As you point out, it probably make sense to have "the state" as a separate thing (in the future, since it's just "a bool" for now). So then the actual check would be "new_state != old_state" (that is not comparing the serialized versions). I still think it makes sense to keep the "thing actually sent" so that a new client gets _exactly_ what the other clients saw.

So maybe all you're getting at is that "when" we have the state abstracted, we should keep "last_state" to compare against? (I agree with that, but I'd still be keen to keep "last_sent" as the thing to give new clients upon connection)

Reply:
> I stored the last "actual thing we sent" to have maximum consistency between clients...what are you suggesting storing, exactly?

I think probably the dict before json encoding it. The serialization should be deterministic, so I don't think consistency is an issue.

> (I'm not sure what you mean by the no-op state updates comment?)

I think most of the calls to `_maybe_update_clients` will not result in sending a message.

Reply:
> the "thing actually sent" so that a new client gets exactly what the other clients saw.

They'll get exactly the same thing even if we serialize again. It feels like it adds complexity to avoid I'm not sure what.

> As you point out, it probably make sense to have "the state" as a separate thing (in the future, since it's just "a bool" for now). 

I think I would go with the dict, even though right now it is equivalent to a bool; I think it gets us close towards the abstraction we want to end up with.